Add boot.efi

wjk opened the pull request:
This PR will add a simple implementation of `boot.efi` using the xnu-loading code from [Apple’s boot-132](https://github.com/apple-oss-distributions/boot/tree/boot-132), modified to function as 64-bit and under EFI.

Note that I currently have no way to test this, because PureDarwin image creation has not yet been implemented.

wjk committed:
Add cctools mtoc to build

wjk committed:
Add gnu-efi-3.0.15 from Debian

This will be used to make building boot.efi much easier.

ninjadev64 commented:
> Note that I currently have no way to test this, because PureDarwin image creation has not yet been implemented.

Naive question: what's wrong with
- Copying all the files from build/src/Kernel/xnu/xnu/ to a volume
- Placing the kexts from build/src/Kernel/Extensions/*/*.kext in System/Library/Extensions/ on the volume
- Of course there'll need to be lots more binaries for libraries and commands we want to have on the system, but that should be enough to test
- Including the bootloader we want to test
- Using `dd` to create a raw image file of the volume and launching it with an OpenCore image as the first drive

RandomDSdevel commented:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;It looks like later released versions of Apple's cctools, including the latest at [v1009.2](https://github.com/apple-oss-distributions/cctools/tree/cctools-1009.2), include x86-64 support; why not use that?

wjk commented:
> Naive question: what's wrong with

The problem is that I would be installing into a DMG file, and there is no way (that I know of) to set the GUID of a partition in a mounted DMG to the required one so it can be considered as a boot partition. I’ve tried `gpt(8)` but it seems Apple has censored the utility; all it does when I run it is say operation not permitted.

ninjadev64 commented:
Ah, I'm not using a DMG file - I added an extra harddisk to my macOS VM and am using that. Would it be possible to set the desired flags on a "physical" volume?

wjk commented:
I’m sorry, I should have been more clear. I am using a raw image as my “DMG file”. macOS doesn’t support loopback mount, but can mount a raw byte-for-byte filesystem image as a DMG. Disk Utility is the best way to create these images, and it outputs files using the .dmg extension. Since this is a raw image, I can easily have QEMU use that as its hard drive when booting the VM.

I would never think of ever letting this code anywhere near a physical hard drive due to the risk of corruption. (Theoretically, wouldn’t you be able to point a command manipulating GPT at a device file and have it perform the same I/O to that device it does when run on an image file?)

ninjadev64 commented:
What I meant was create the image on a "physical" hard drive, which isn't really physical since I use macOS in a VM, then create an image of that using dd to use with QEMU inside the macOS VM.

wjk committed:
Set output name for libefi

wjk committed:
Add required leading underscore to symbol names

RandomDSdevel commented:
(Replying belatedly:)  

> …macOS doesn’t support loopback mount, …

 1. macOS actually did have a loop(back) device 'driver' as recently as 10.14.6, if you weren't aware, if I'm understanding the correlation between macOS and XNU version numbers correctly (though maybe only in development kernel builds.)  Refer to ['`bsd/dev/vn`'](https://github.com/apple-oss-distributions/xnu/tree/xnu-8020.140.41/bsd/dev/vn) in [XNU's 8020.140.41 sources](https://github.com/apple-oss-distributions/xnu/tree/xnu-8020.140.41/) for its implementation as it last appeared before getting removed.  (You'll want to make special note of note 2 at the top of '[`bad/dev/vn/vn.c`'](https://github.com/apple-oss-distributions/xnu/blob/xnu-8020.140.41/bsd/dev/vn/vn.c),' however.)  
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;You could maybe start with that and port it forward.  That only helps once PureDarwin can be booted, though, of course.  Or perhaps/maybe if onew were really lucky, unlikely as that might be, that implementation hadn't diverged too far from FreeBSD's at the time and the current upstream implementation can just get merged back in again.  (Though it might well be rather surprising if that ended up being the case.)  
 1. macFUSE has a loop(back) filesystem implementation in [LoopbackFS](https://github.com/osxfuse/filesystems/tree/master/filesystems-c/loopback), but it's currently un-maintained.  That also only helps from the macOS side.

wjk committed:
Initial commit of boot.efi

hideout commented:
Hey you are right   LoopbackFS
<https://github.com/osxfuse/filesystems/tree/master/filesystems-c/loopback>,
is totally outdated last commit date 5 years ago

anyway thanks for the work

Am Sa., 11. Mai 2024 um 22:40 Uhr schrieb Bryce Glover <
***@***.***>:

> (Replying belatedly:)
>
> …macOS doesn’t support loopback mount, …
>
>
>    1. macOS actually did have a loop(back) device 'driver' as recently as
>    10.14.6, if you weren't aware, if I'm understanding the correlation between
>    macOS and XNU version numbers correctly (though maybe only in development
>    kernel builds.) Refer to 'bsd/dev/vn'
>    <https://github.com/apple-oss-distributions/xnu/tree/xnu-8020.140.41/bsd/dev/vn>
>    in XNU's 8020.140.41 sources
>    <https://github.com/apple-oss-distributions/xnu/tree/xnu-8020.140.41/>
>    for its implementation as it last appeared before getting removed. (You'll
>    want to make special note of note 2 at the top of 'bad/dev/vn/vn.c'
>    <https://github.com/apple-oss-distributions/xnu/blob/xnu-8020.140.41/bsd/dev/vn/vn.c>,'
>    however.)
>         You could maybe start with that and port it forward. That only
>    helps once PureDarwin can be booted, though, of course. Or perhaps/maybe if
>    onew were really lucky, unlikely as that might be, that implementation
>    hadn't diverged too far from FreeBSD's at the time and the current upstream
>    implementation can just get merged back in again. (Though it might well be
>    rather surprising if that ended up being the case.)
>    2. macFUSE has a loop(back) filesystem implementation in LoopbackFS
>    <https://github.com/osxfuse/filesystems/tree/master/filesystems-c/loopback>,
>    but it's currently un-maintained. That also only helps from the macOS side.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PureDarwin/PureDarwin/pull/138#issuecomment-2106026015>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEMNDMA4NOEJBYNPDFJNYLZBZ63BAVCNFSM6AAAAABBRTGPEWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMBWGAZDMMBRGU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

wjk commented:
If anyone can help me out here, I am stumped on something.

I am trying to locate the `EFI_FILE_PROTOCOL` for the root of the volume on which the running boot.efi is located, so I can locate and load `com.apple.Boot.plist` and the boot kexts. I’ve read the UEFI spec, as well as the source code for the GRUB and rEFIt bootloaders, and I can’t come up with anything similar. However, the only parameter I have to work with is the ImageHandle parameter passed to the EFI entry point. Does anyone have any pointers?

wjk committed:
Add util.c to boot.efi

This header contains a definition for NULL, as well as
a copy of the strcmp() and strncmp() functions. It will
be extended to contain an EFI-compatible malloc()
implementation later.

wjk committed:
Add dead simple malloc/free implementation to util.c

ninjadev64 commented:
I don't know about that myself, but your question has been forwarded on to the Discord server by csekel so you might get a response soon (just to let you know your question has been noticed).

wjk committed:
Add strlen() implementation

wjk committed:
Add panic() implementation

wjk committed:
Add strcpy() implementation

wjk committed:
Add plist.c and matching header

These functions come from the boot-132 file i386/libsaio/xml.c.
Symbols have been renamed and moved around to better clean up
the source code.

wjk committed:
Add bzero() function

wjk committed:
Add bcopy() implementation

wjk committed:
Avoid redefinition of NULL

wjk committed:
Add device_tree.c to boot.efi target

These functions come from bopt-132 i386/libsaio/device_tree.c.
Some symbols have been renamed for cleanup purposes.

wjk commented:
@ninjadev64 Small update for you. Unfortunately, it doesn’t look like EFI is able to parse HFS+ filesystems natively, so `EFI_FILE_PROTOCOL` won’t work. As a result, I need to obtain the device that boot.efi is located on instead, so I can use the EFI volume I/O protocols. Thanks!

wjk committed:
Add extra newline in error string

RandomDSdevel commented:
(Continuing on from [my earlier comment](https://github.com/PureDarwin/PureDarwin/pull/138#issuecomment-2106026015) in earlier discussion of loop(back) mounts/filesystems:)  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Also, the [FSKit](https://developer.apple.com/documentation/fskit) user-space/user-mode filesystem API framework will be made available for public use as of macOS Sequoia (v15) (and its accompanying backup XNU kernel version) this fall.

csekel commented:
> (Continuing on from [my earlier comment](https://github.com/PureDarwin/PureDarwin/pull/138#issuecomment-2106026015) in earlier discussion of loop(back) mounts/filesystems:)
> 
>      Also, the [FSKit](https://developer.apple.com/documentation/fskit) user-space/user-mode filesystem API framework will be made available for public use as of macOS Sequoia (v15) (and its accompanying backup XNU kernel version) this fall.

Let’s hope that they open source it

Bendy5410 commented:
> If anyone can help me out here, I am stumped on something.
> 
> I am trying to locate the `EFI_FILE_PROTOCOL` for the root of the volume on which the running boot.efi is located, so I can locate and load `com.apple.Boot.plist` and the boot kexts. I’ve read the UEFI spec, as well as the source code for the GRUB and rEFIt bootloaders, and I can’t come up with anything similar. However, the only parameter I have to work with is the ImageHandle parameter passed to the EFI entry point. Does anyone have any pointers?

Have you solved this problem? If not, it may come in handy: https://github.com/acidanthera/OpenCorePkg/tree/master/Staging/OpenHfsPlus.

AS5410-developer commented:
> @ninjadev64 Small update for you. Unfortunately, it doesn’t look like EFI is able to parse HFS+ filesystems natively, so `EFI_FILE_PROTOCOL` won’t work. As a result, I need to obtain the device that boot.efi is located on instead, so I can use the EFI volume I/O protocols. Thanks!

Have you solved this problem? If not, it may come in handy.: https://github.com/acidanthera/OpenCorePkg/tree/master/Staging/OpenHfsPlus.

wjk commented:
@AS5410-developer Looks useful, but it still doesn’t fix my main problem: how to determine which EFI volume the currently executing copy of `boot.efi` lives on. It will take quite a bit of time to separate that code from the rest of OpenCore and convert it to gnu-efi, however.

AS5410-developer commented:
> how to determine which EFI volume the currently executing copy of `boot.efi` lives on.

Most often, a config file with a clearly defined partition and disk is created on the OS installed on the hard disk. Ramdisk is used on installation disks. I think it would be possible to create a config file with the file system id and a unique id (maybe 512 bytes to minimize the likelihood of a match if 2-3 systems are installed on different disks). You can make some kind of volume label. But don't take me seriously. I'm imperfectly summer, and I didn't go to college or university.

AS5410-developer commented:
> how to determine which EFI volume the currently executing copy of `boot.efi` lives on.

Doesn't the boot argument "rd" tell the root device?

AS5410-developer commented:
> It will take quite a bit of time to separate that code from the rest of OpenCore and convert it to gnu-efi, however.

Doesn't it stay loaded as a UEFI driver after running another efi program?

wjk committed:
Fixes regarding EFI function calling conventions
